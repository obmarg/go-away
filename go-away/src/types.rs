//! Defines the type model for go-away - a set of structs that describe
//! types and can be used to generate copies of them in other languages.

/// A struct.
///
/// This will be serialized as a JSON object.
#[derive(Debug)]
pub struct Struct {
    /// The name of the struct in Rust
    pub name: String,

    /// The structs fields.
    pub fields: Vec<Field>,
}

/// A field within a struct
#[derive(Debug)]
pub struct Field {
    /// The name of the field in rust.  If the field is un-named this will
    /// be a number.
    pub name: String,
    // TODO: Need both serialize & deserialize apparently
    /// The name of the field when serialized.
    pub serialized_name: String,
    /// The type of the field
    pub ty: FieldType,
}

/// A newtype struct (e.g. `struct SomeId(String)`)
///
/// These are usually represented as their inner type when serialized.
#[derive(Debug)]
pub struct NewType {
    /// The name of the struct in rust.
    pub name: String,

    /// The type contained within the newtype.
    pub inner: FieldType,
}

/// A type alias (e.g. `type SomeType = HashMap<String, String>;`)
///
/// These are usually represented as their inner type when serialized.
#[derive(Debug)]
pub struct Alias {
    /// The name of the type alias in rust.
    pub name: String,

    /// The type that is being aliased.
    pub inner: FieldType,
}

/// An enum - note that in go-away these do not contain data.
///
/// A Rust enum that's variants contain values will go to a `UnionType`
#[derive(Debug)]
pub struct Enum {
    /// The name of the enum
    pub name: String,
    /// The enums variants
    pub variants: Vec<EnumVariant>,
}

/// An enum variant - note that these are just names and are serialized
/// as strings.
#[derive(Debug)]
pub struct EnumVariant {
    /// The name of the variant in code.
    pub name: String,
    /// The name of the variant when serialized.
    pub serialized_name: String,
}

/// A union type - any rust enum that's variants contain data.
///
/// These will be serialzied differently depending on the UnionRepresentation.
#[derive(Debug)]
pub struct Union {
    /// The name of the union
    pub name: String,
    /// The representation of the union to use on the wire.
    pub representation: UnionRepresentation,
    /// The unions variants.
    pub variants: Vec<UnionVariant>,
}

#[derive(Debug, PartialEq, Eq)]
/// A variant of a union type
pub struct UnionVariant {
    /// The name of the variant if any
    pub name: Option<String>,
    /// The type inside the variant
    pub ty: FieldType,
    /// The name the variant will be serialized to
    pub serialized_name: String,
}

/// The serialized representation of the union type
///
/// See https://serde.rs/enum-representations.html for details
#[derive(Debug)]
pub enum UnionRepresentation {
    /// An adjacently tagged representation
    AdjacentlyTagged {
        /// The name of the tag field
        tag: String,
        /// The name of the content field
        content: String,
    },
    /// An internally tagged representation
    InternallyTagged {
        /// The name of the tag field
        tag: String,
    },
    /// An externally tagged representation
    ExternallyTagged,
    /// An untagged representation
    Untagged,
}

/// The type of a field.
#[derive(Debug, PartialEq, Eq)]
pub enum FieldType {
    /// A `Option<T>` field
    Optional(Box<FieldType>),
    /// a `Vec<T>` field
    List(Box<FieldType>),
    /// a `HashMap<K, V>` field
    Map {
        /// The type of the HashMaps keys
        key: Box<FieldType>,
        /// The type of the HashMaps values
        value: Box<FieldType>,
    },
    /// A field with a named type
    Named(TypeRef),
    /// A field with a primitive type
    Primitive(Primitive),
}

/// The primitive types
#[derive(Debug, PartialEq, Eq)]
pub enum Primitive {
    /// Strings
    String,
    /// Floating point numbers
    Float,
    /// Integers
    Int,
    /// Booleans
    Bool,
    /// Time
    Time,
}

/// A reference to a given named type
#[derive(Debug, PartialEq, Eq)]
pub struct TypeRef {
    pub(crate) name: String,
    // TODO: id: std::any::TypeId,
}

impl TypeRef {
    /// Gets the name of the referenced type
    pub(crate) fn name(&self) -> &str {
        &self.name
    }
}
