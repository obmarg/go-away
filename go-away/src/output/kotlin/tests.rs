use insta::assert_snapshot;

use super::*;
use crate::types::{EnumVariant, Primitive, TypeRef, UnionRepresentation, UnionVariant};

#[test]
fn test_primitive_structs() {
    assert_snapshot!(
        KotlinType::Struct(&Struct {
            name: "MyStruct".into(),
            fields: vec![
                Field {
                    name: "a_string".into(),
                    serialized_name: "a_string".into(),
                    ty: FieldType::Primitive(Primitive::String),
                },
                Field {
                    name: "an_int".into(),
                    serialized_name: "renamed_tho".into(),
                    ty: FieldType::Primitive(Primitive::Int),
                },
                Field {
                    name: "a_bool".into(),
                    serialized_name: "also_renamed".into(),
                    ty: FieldType::Primitive(Primitive::Bool),
                },
                Field {
                    name: "a_float".into(),
                    serialized_name: "a_float".into(),
                    ty: FieldType::Primitive(Primitive::Float),
                },
                Field {
                    name: "optionalFloat".into(),
                    serialized_name: "a_float".into(),
                    ty: FieldType::Optional(Box::new(FieldType::Primitive(Primitive::Float))),
                },
            ],
        })
        .to_string(),
        @r###"
    @Serializable
    data class MyStruct(
        @SerialName("a_string")
        public var aString: String,
        @SerialName("renamed_tho")
        public var anInt: Long,
        @SerialName("also_renamed")
        public var aBool: Boolean,
        @SerialName("a_float")
        public var aFloat: Double,
        @SerialName("a_float")
        public var optionalFloat: Double? = null,
    )


    "###
    );
}

#[test]
fn test_newtype_output() {
    assert_snapshot!(KotlinType::NewType(&NewType {
            name: "UserId".into(),
            inner: FieldType::Primitive(Primitive::String),
        })
        .to_string(), @r###"
    @Serializable(with = UserIdSerializer::class)
    data class UserId(
        @SerialName("value")
        public var value: String,
    )


    object UserIdSerializer : KSerializer<UserId> {
        private val serializer = String.serializer();

        override val descriptor: SerialDescriptor = serializer.descriptor;

        override fun serialize(encoder: Encoder, value: UserId) {
            encoder.encodeSerializableValue(serializer, value.value)
        }

        override fun deserialize(decoder: Decoder): UserId {
            return UserId(decoder.decodeSerializableValue(serializer))
        }
    }

    "###);
}

#[test]
fn test_enum_output() {
    assert_snapshot!(KotlinType::Enum(&types::Enum {
            name: "FulfilmentType".into(),
            variants: vec![
                EnumVariant {
                    name: "Delivery".into(),
                    serialized_name: "DELIVERY".into(),
                },
                EnumVariant {
                    name: "Collection".into(),
                    serialized_name: "COLLECTION".into(),
                },
            ],
        })
        .to_string(), @r###"
    @Serializable
    public enum class FulfilmentType {
        @SerialName("DELIVERY") DELIVERY,
        @SerialName("COLLECTION") COLLECTION,
    }


    "###);
}

#[test]
fn test_adjacently_tagged_union_output() {
    assert_snapshot!(KotlinType::Union(&types::Union {
        name: "MyUnion".into(),
        representation: UnionRepresentation::AdjacentlyTagged {
            tag: "type".into(),
            content: "data".into(),
        },
        variants: vec![
            UnionVariant {
                name: Some("VarOne".into()),
                ty: FieldType::Named(TypeRef {
                    name: "VarOne".into()
                }),
                serialized_name: "VAR_ONE".into(),
            },
            UnionVariant {
                name: Some("VarTwo".into()),
                ty: FieldType::Named(TypeRef {
                    name: "VarTwo".into()
                }),
                serialized_name: "VAR_TWO".into(),
            }
        ]
    })
    .to_string());
}

#[test]
fn test_list_types() {
    assert_snapshot!(
        FieldType::List(Box::new(FieldType::Primitive(Primitive::String))).kotlin_type(),
        @"List<String>"
    );
}

#[test]
fn test_map_types() {
    assert_snapshot!(
        FieldType::Map{
            key: Box::new(FieldType::Primitive(Primitive::String)),
            value: Box::new(FieldType::Primitive(Primitive::Int))
        }.kotlin_type(),
        @"Map<String, Long>"
    );
}

#[test]
fn test_option_types() {
    assert_snapshot!(
        FieldType::Optional(Box::new(FieldType::Primitive(Primitive::String))).kotlin_type(),
        @"String?"
    );
}
