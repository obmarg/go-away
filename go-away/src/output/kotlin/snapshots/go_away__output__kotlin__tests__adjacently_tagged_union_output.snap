---
source: go-away/src/output/kotlin/tests.rs
expression: "KotlinType::Union(&types::Union {\n                name: \"MyUnion\".into(),\n                representation: UnionRepresentation::AdjacentlyTagged {\n                    tag: \"type\".into(),\n                    content: \"data\".into(),\n                },\n                variants: vec![UnionVariant\n                    {\n                        name : Some(\"VarOne\".into()), ty : FieldType ::\n                        Named(TypeRef { name : \"VarOne\".into() }), serialized_name :\n                        \"VAR_ONE\".into(),\n                    }, UnionVariant\n                    {\n                        name : Some(\"VarTwo\".into()), ty : FieldType ::\n                        Named(TypeRef { name : \"VarTwo\".into() }), serialized_name :\n                        \"VAR_TWO\".into(),\n                    }],\n            }).to_string()"
---
@Serializable(with = MyUnionSerializer::class)
sealed interface MyUnion {
    @Serializable(with = varOneSerializer::class)
    data class varOne: MyUnion(
        @SerialName("value")
        public var value: VarOne,
    )


    object varOneSerializer : KSerializer<varOne> {
        private val serializer = VarOne.serializer();

        override val descriptor: SerialDescriptor = serializer.descriptor;

        override fun serialize(encoder: Encoder, value: varOne) {
            encoder.serialize(serializer, value.value)
        }

        override fun deserialize(decoder: Decoder): varOne {
            return varOne(decoder.decode(serializer))
        }
    }

    @Serializable(with = varTwoSerializer::class)
    data class varTwo: MyUnion(
        @SerialName("value")
        public var value: VarTwo,
    )


    object varTwoSerializer : KSerializer<varTwo> {
        private val serializer = VarTwo.serializer();

        override val descriptor: SerialDescriptor = serializer.descriptor;

        override fun serialize(encoder: Encoder, value: varTwo) {
            encoder.serialize(serializer, value.value)
        }

        override fun deserialize(decoder: Decoder): varTwo {
            return varTwo(decoder.decode(serializer))
        }
    }

}


